linera-core/src/client/mod.rs

```diff
@@ -1138,78 +1144,106 @@ where
         // Retrieve newly received certificates from this validator.
         let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_n(tracker);
         let info = remote_node.handle_chain_info_query(query).await?;
-        let mut certificates = Vec::new();
+        let mut certificates: Vec<Certificate> = Vec::new();
         let mut new_tracker = tracker;
+
         for entry in info.requested_received_log {
             let query = ChainInfoQuery::new(entry.chain_id)
                 .with_sent_certificate_hashes_in_range(BlockHeightRange::single(entry.height));
+
             let local_response = self
                 .client
                 .local_node
                 .handle_chain_info_query(query.clone())
                 .await
                 .map_err(|error| NodeError::LocalNodeQuery {
                     error: error.to_string(),
-                })?;
-            if !local_response
-                .info
-                .requested_sent_certificate_hashes
-                .is_empty()
-            {
+                })?
+                .info;
+            if !local_response.requested_sent_certificate_hashes.is_empty() {
+                // We've already processed incoming messages for this certificate.
+                // TODO: We verify it's non-empty but how do we make sure we've processed ALL?
```

Comment:
@afck can you have a look at this question? I left it in the code so I don't forget.

Reply:
We only requested a single entry, so the collection contains either that or nothing.